velocity/src/main/java/com/github/games647/fastlogin/velocity/task/ForceLoginTask.java

```diff
@@ -0,0 +1,141 @@
+
+    //treat player as if they had a premium account, even when they don't
+    //used for Floodgate auto login/register
+    private final boolean forcedOnlineMode;
+
+    public ForceLoginTask(FastLoginCore<Player, CommandSource, FastLoginVelocity> core,
+                          Player player, RegisteredServer server, VelocityLoginSession session, boolean forcedOnlineMode) {
+        super(core, player, session);
+
+        this.server = server;
+        this.forcedOnlineMode = forcedOnlineMode;
+    }
+
+    public ForceLoginTask(FastLoginCore<Player, CommandSource, FastLoginVelocity> core, Player player,
+                          RegisteredServer server, VelocityLoginSession session) {
+        this(core, player, server, session, false);
+    }
+
+    @Override
+    public void run() {
+        if (session == null) {
+            return;
+        }
+
+        super.run();
+        if (!isOnlineMode()) {
+            session.setAlreadySaved(true);
+        }
+    }
+
+    @Override
+    public boolean forceLogin(Player player) {
+        if (session.isAlreadyLogged()) {
+            return true;
+        }
+
+        session.setAlreadyLogged(true);
+        return super.forceLogin(player);
+    }
+
+    @Override
+    public FastLoginAutoLoginEvent callFastLoginAutoLoginEvent(LoginSession session, StoredProfile profile) {
+        //FIXME: Am I doing it right?
+        VelocityFastLoginAutoLoginEvent event = new VelocityFastLoginAutoLoginEvent(session, profile);
+        try {
+             return core.getPlugin().getProxy().getEventManager().fire(event).get();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt(); // Set the interrupt flag again
+            return event;
+        } catch (ExecutionException e) {
+            e.printStackTrace();
+            return event;
+        }
+    }
+
+    @Override
+    public boolean forceRegister(Player player) {
+        return session.isAlreadyLogged() || super.forceRegister(player);
+    }
+
+    @Override
+    public void onForceActionSuccess(LoginSession session) {
+        //sub channel name
+        Type type = Type.LOGIN;
+        if (session.needsRegistration()) {
+            type = Type.REGISTER;
+        }
+        //FIXME: Velocity does not have an alternative for this!
+        //UUID proxyId = UUID.fromString(ProxyServer.getInstance().getConfig().getUuid());
+        UUID proxyId = UUID.nameUUIDFromBytes("velocity".getBytes(StandardCharsets.UTF_8));
+        ChannelMessage loginMessage = new LoginActionMessage(type, player.getUsername(), proxyId);
+
+        core.getPlugin().sendPluginMessage(server, loginMessage);
+    }
+
+    @Override
+    public String getName(Player player) {
+        return player.getUsername();
+    }
+
+    @Override
+    public boolean isOnline(Player player) {
+        //FIXME: is this right?
+        return core.getPlugin().getProxy().getPlayer(player.getUsername()).isPresent();
```

Comment:
This could end up wrong. If there is another login request is coming in, this would evaluate to true even though it's not the same connection or even person.

Reply:
I missed that there was an equivalent for `#isConnected()`, my bad.